File: apps/marlin-wallet/android/app/src/main/java/com/marlinwallet/app/WearableListenerService.java
// Copyright (c) 2025 The Bitcoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

package com.marlinwallet.app;

import android.content.Context;
import android.content.Intent;
import com.google.android.gms.wearable.DataClient;
import com.google.android.gms.wearable.DataEvent;
import com.google.android.gms.wearable.DataEventBuffer;
import com.google.android.gms.wearable.DataItem;
import com.google.android.gms.wearable.DataMap;
import com.google.android.gms.wearable.DataMapItem;
import com.google.android.gms.wearable.MessageEvent;
import com.google.android.gms.wearable.PutDataMapRequest;
import com.google.android.gms.wearable.PutDataRequest;
import com.google.android.gms.wearable.Wearable;

/**
 * Service to handle communication with Wear OS watch
 */
public class WearableListenerService extends com.google.android.gms.wearable.WearableListenerService {
    
    private static final String WALLET_DATA_PATH = "/wallet_data";
    private static final String KEY_ADDRESS = "address";
    private static final String KEY_BIP21_PREFIX = "bip21_prefix";
    private static final String OPEN_APP_MESSAGE = "/open_app";
    
    @Override
    public void onMessageReceived(MessageEvent messageEvent) {
        if (messageEvent.getPath().equals(OPEN_APP_MESSAGE)) {
            // Watch is requesting to open the phone app
            Intent intent = new Intent(this, MainActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(intent);
        }
    }
    
    /**
     * Send wallet data (address and BIP21 prefix) to watch
     * This is called from React Native when the wallet is ready
     */
    public static void sendDataToWatch(Context context, String address, String bip21Prefix) {
        if (address == null || address.isEmpty()) {
            return;
        }
        
        // Create a data map with the address and BIP21 prefix
        PutDataMapRequest dataMap = PutDataMapRequest.create(WALLET_DATA_PATH);
        dataMap.getDataMap().putString(KEY_ADDRESS, address);
        dataMap.getDataMap().putString(KEY_BIP21_PREFIX, bip21Prefix != null ? bip21Prefix : "ecash:");
        dataMap.getDataMap().putLong("timestamp", System.currentTimeMillis()); // Force update
        
        PutDataRequest request = dataMap.asPutDataRequest();
        request.setUrgent();
        
        // Send to watch
        DataClient dataClient = Wearable.getDataClient(context);
        dataClient.putDataItem(request);
    }
}

File: apps/marlin-wallet/android/wear/src/main/java/com/marlinwallet/app/wear/WearableListenerService.kt
// Copyright (c) 2025 The Bitcoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

package com.marlinwallet.app.wear

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import com.google.android.gms.wearable.DataEvent
import com.google.android.gms.wearable.DataEventBuffer
import com.google.android.gms.wearable.DataMapItem
import com.google.android.gms.wearable.WearableListenerService

class WearableListenerService : WearableListenerService() {
    
    companion object {
        private const val WALLET_DATA_PATH = "/wallet_data"
        private const val KEY_ADDRESS = "address"
        private const val KEY_BIP21_PREFIX = "bip21_prefix"
    }
    
    override fun onDataChanged(dataEvents: DataEventBuffer) {
        for (event in dataEvents) {
            if (event.type == DataEvent.TYPE_CHANGED) {
                val item = event.dataItem
                if (item.uri.path == WALLET_DATA_PATH) {
                    // Extract address and BIP21 prefix from data
                    val dataMap = DataMapItem.fromDataItem(item).dataMap
                    val address = dataMap.getString(KEY_ADDRESS)
                    val bip21Prefix = dataMap.getString(KEY_BIP21_PREFIX)
                    
                    if (address != null) {
                        // Check if this is the first time receiving address
                        val prefs = getSharedPreferences("wallet", MODE_PRIVATE)
                        val previousAddress = prefs.getString("address", null)
                        val wasInitialized = previousAddress != null
                        
                        // Save both address and BIP21 prefix to SharedPreferences
                        val editor = prefs.edit()
                        editor.putString("address", address)
                        if (bip21Prefix != null) {
                            editor.putString("bip21_prefix", bip21Prefix)
                        }
                        editor.apply()
                        
                        // Only start/refresh MainActivity if needed
                        if (!wasInitialized) {
                            // First initialization - start the app fresh
                            val intent = Intent(this, MainActivity::class.java)
                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
                            startActivity(intent)
                        } else if (isAppInForeground()) {
                            // App is already running - refresh it without bringing to foreground
                            val intent = Intent(this, MainActivity::class.java)
                            intent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
                            startActivity(intent)
                        }
                        // If app is initialized but not in foreground, do nothing
                    }
                }
            }
        }
    }
    
    private fun isAppInForeground(): Boolean {
        val activityManager = getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val appProcesses = activityManager.runningAppProcesses ?: return false
        
        for (appProcess in appProcesses) {
            if (appProcess.processName == packageName) {
                return appProcess.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND
            }
        }
        return false
    }
}

